pochivm/api_base.h

```diff
@@ -115,113 +116,186 @@ Value<T>::operator Value<U>() const
     return StaticCast<U>(*this);
 }
 
+// Language utility: construct a literal
+// Example: Literal<int64_t>(1)
+//
+template<typename T>
+Value<T> Literal(T x)
+{
+    return Value<T>(new AstLiteralExpr(TypeId::Get<T>(), &x));
+}
+
+// If src is a literal expression, return a new equivalent literal expression of type U. Otherwise, return
+// a StaticCast of src to U
+//
+template<typename U, typename T, typename = std::enable_if_t<AstTypeHelper::may_static_cast<T, U>::value> >
+Value<U> StaticCastOrConvertLiteral(const Value<T>& src)
+{
+    static_assert(AstTypeHelper::may_static_cast<T, U>::value, "cannot static_cast T to U");
+    if(src.__pochivm_value_ptr->GetAstNodeType() == AstNodeType::AstLiteralExpr)
+    {
+        return Literal<U>(static_cast<U>(reinterpret_cast<AstLiteralExpr *>(src.__pochivm_value_ptr)->GetAs<T>()));
```

Comment:
please go to cppreference.com and learn the different types of CPP casts. Specifically, reinterpret_cast is only used to reinterpret a pointer as an integer or vice versa.
The type of cast you want to do here is a downcast, so you should use either static_cast (silently breaks down if the type is wrong) or dynamic_cast (gracefully provides a runtime error, but incurs a large runtime cost). In our system this is addressed by our assert_cast helper, which does dynamic_cast in debug build but static_cast in non-debug build.

Reply:
as a side note, using `reinterpret_cast` to do upcast or downcast is not a style issue but a a correctness issue because it breaks down if the cast results in a pointer shift, see example below:
```
struct A { int x; };
struct B { int y; };
struct C : public A, B {};
C* c = new C();
c->x = 1; c->y = 2;
B* b1 = static_cast<B*>(c);
printf("%d\n", b1->y);  // gives you 2 correctly
B* b2 = reinterpret_cast<B*>(c);
printf("%d\n", b2->y);  // gives you 1 instead of 2
```

Reply:
Sorry my bad. That was a stupid mistake